src/controllerclientimpl.cpp

```diff
@@ -475,6 +484,38 @@ void ControllerClientImpl::ExecuteGraphQueryRaw(const char* operationName, const
     _ExecuteGraphQuery(operationName, query, rVariables, rResult, rAlloc, timeout, false, true);
 }
 
+void ControllerClientImpl::ExecuteGraphSubscription(const char* operationName, const char* query, const rapidjson::Value& rVariables, rapidjson::Document::AllocatorType& rAlloc) 
+{
+    boost::mutex::scoped_lock lock(_mutex);
+    if (!_pWebsocketStream->is_open()) {
+        CreateWebSocketStream();
+    }
+    rapidjson::StringBuffer& rRequestStringBuffer = _rRequestStringBufferCache;
+    rRequestStringBuffer.Clear();
+
+    rapidjson::Value rRequest, rPayLoad, rValue;
+    rRequest.SetObject();
+    rPayLoad.SetObject();
+    rValue.SetString(operationName, rAlloc);
+    rPayLoad.AddMember(rapidjson::Document::StringRefType("operationName"), rValue, rAlloc);
+    rValue.SetString(query, rAlloc);
+    rPayLoad.AddMember(rapidjson::Document::StringRefType("query"), rValue, rAlloc);
+    rValue.CopyFrom(rVariables, rAlloc);
+    rPayLoad.AddMember(rapidjson::Document::StringRefType("variables"), rValue, rAlloc);
+    rValue.SetString("start", rAlloc);
+    rRequest.AddMember(rapidjson::Document::StringRefType("type"), rValue, rAlloc);
+    rRequest.AddMember(rapidjson::Document::StringRefType("payload"), rPayLoad, rAlloc);
+
+    rapidjson::Writer<rapidjson::StringBuffer> writer(rRequestStringBuffer);
+    rRequest.Accept(writer);
+    _pWebsocketStream->write(rRequestStringBuffer.GetString());
+    rRequestStringBuffer.Clear();
+}
+
+std::string ControllerClientImpl::SpinOnce(){
```

Comment:
Not sure whether we should process the read message on the client side:
1. Parse the message with RapidJSON .
2. Skip the keep-alive message {"type":"ka"} and continue reading until we receive a valid message.

Reply:
If it is the user's job to call `SpinOce`, then you don't need the `SpinOnce` function. `SpinOnce` is usually used when there is a background thread. In this case, maybe you can consider to have a function `GetSubriptionResult` or something similar to let the use to fetch the data when they need it.